### PR TITLE
push client the new root cert when it's changed

### PR DESCRIPTION
**Pull request by quanjielin**
asked from - https://github.com/istio/istio/issues/13158

### CODE REVIEW

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -226,6 +230,25 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 	return ns, nil
 }
 
+// SecretExist checks if secret already existed.
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: The comment does not provide too much info. Suggest to change to something that describes why we need this function.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -379,41 +412,41 @@ func (sc *SecretCache) rotate() {
...
+			// Send the notification to close the stream if token is expired, so that client could re-connect with a new token.
+			if sc.isTokenExpired() {
+				log.Debugf("Token for %q expired for proxy %q", resourceName, connectionID)
+
+				if sc.notifyCallback != nil {
+					if err := sc.notifyCallback(connectionID, resourceName, nil /*nil indicates close the streaming connection to proxy*/); err != nil {
+						log.Errorf("Failed to notify for proxy %q: %v", connectionID, err)
+					}
+				} else {
+					log.Warnf("secret cache notify callback isn't set")
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: s/Warnf/Warn

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -533,7 +566,7 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
 	}
 
 	// Cert exipre time by default is createTime + sc.configOptions.SecretTTL.
-	// Current CAs(Citadel, GoogleCA) respect SecretTTL that passed to it and use it decide TTL of cert it issued.
+	// Current CAs(Citadel, etc) respect SecretTTL that passed to it and use it decide TTL of cert it issued.
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: "Citadel respects SecretTTL..."

File: security/pkg/nodeagent/cache/secretcache.go [hunk 4]
```diff
@@ -552,12 +585,16 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+	rootCertChanged := !bytes.Equal(sc.rootCert, []byte(certChainPEM[length-1]))
+	if sc.rootCert == nil || rootCertChanged {
 		sc.rootCertMutex.Lock()
 		sc.rootCert = []byte(certChainPEM[length-1])
 		sc.rootCertMutex.Unlock()
+	}
 
-		// TODO(quanlin): notify proxy about root cert change.
+	if rootCertChanged {
+		log.Infof("Root cert has changed")
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: s/Infof/Info

File: security/pkg/nodeagent/cache/secretcache.go (hunk 3 above)
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: The generateSecret function has more than 100 lines. Consider breaking it into small pieces for readability concern. Can be in a separate PR

Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: resolve other comments, will do this one in separate PR

File: security/pkg/nodeagent/cache/secretcache.go [hunk 5]
```diff
@@ -226,6 +231,26 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
...
+		ConnectionID: connectionID,
+		ResourceName: resourceName,
+	}
+	val, exist := sc.secrets.Load(key)
+	if !exist {
+		return false
+	}
+
+	e := val.(model.SecretItem)
+	if e.ResourceName == resourceName && e.Token == token && e.Version == version {
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: You don't need this if block, just 

return e.ResourceName == resourceName && e.Token == token && e.Version == version

File: security/pkg/nodeagent/cache/secretcache.go [hunk 6]
```diff
@@ -226,6 +231,26 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 	return ns, nil
 }
 
+// SecretExist checks if secret already existed.
+// This API is used for sds server to check if coming request is ack request.
+func (sc *SecretCache) SecretExist(connectionID, resourceName, token, version string) bool {
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: This is not a new function and the logic is quite similar. Move this back to the original place? It will be much easier for code review.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 7]
```diff
@@ -235,6 +260,26 @@ func (sc *SecretCache) DeleteSecret(connectionID, resourceName string) {
 	sc.secrets.Delete(key)
 }
 
+// Close shuts down the secret cache.
+func (sc *SecretCache) Close() {
+	sc.closing <- true
+}
+
+func (sc *SecretCache) keyCertRotationJob() {
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Same here, this is not a new function, suggest to move it to the original place for better code review purpose.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 6 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: yes this PR just move this function around, prefer to move public APIs(like ```SecretExist``` to beginning of the file, and push helper function like ```keyCertRotationJob``` lower.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 5 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: done
